File: ios/HybridContinuedTask.swift
import BackgroundTasks
import Foundation
import NitroModules

/// A live handle to one `BGContinuedProcessingTask`.
///
/// The handle is created at submit time and starts in the `pending` state. The
/// system's launch handler fires some time later and calls `attach`, which is
/// the moment the task is really running.
///
/// Mutable state is guarded by a narrow lock because JS reads the properties
/// synchronously on the JS thread while `BGTaskScheduler` invokes the
/// expiration handler on a queue of its choosing. The lock is never held while
/// a listener is invoked.
@available(iOS 26.0, *)
final class HybridContinuedTask: HybridContinuedTaskSpec {
  private let lock = NSLock()

  private let identifier: String
  private var currentTitle: String
  private var currentSubtitle: String
  private var currentState: ContinuedTaskState = .pending
  private var currentCompletedUnitCount: Double = 0
  private var currentTotalUnitCount: Double

  private var nativeTask: BGContinuedProcessingTask?
  private var startListeners: [Int: () -> Void] = [:]
  private var stopListeners: [Int: (TaskStopEvent) -> Void] = [:]
  private var nextListenerId = 0

  init(identifier: String, title: String, subtitle: String, totalUnitCount: Double) {
    self.identifier = identifier
    self.currentTitle = title
    self.currentSubtitle = subtitle
    self.currentTotalUnitCount = totalUnitCount
  }

  // MARK: - Spec properties

  var id: String { identifier }

  var title: String {
    lock.lock()
    defer { lock.unlock() }
    return currentTitle
  }

  var subtitle: String {
    lock.lock()
    defer { lock.unlock() }
    return currentSubtitle
  }

  var state: ContinuedTaskState {
    lock.lock()
    defer { lock.unlock() }
    return currentState
  }

  var completedUnitCount: Double {
    lock.lock()
    defer { lock.unlock() }
    return currentCompletedUnitCount
  }

  var totalUnitCount: Double {
    lock.lock()
    defer { lock.unlock() }
    return currentTotalUnitCount
  }

  // MARK: - Spec methods

  func updateTitle(title: String, subtitle: String) throws {
    lock.lock()
    currentTitle = title
    currentSubtitle = subtitle
    let task = nativeTask
    lock.unlock()

    task?.updateTitle(title, subtitle: subtitle)
    ContinuedTaskStore.shared.update(id: identifier) { record in
      record.title = title
      record.subtitle = subtitle
    }
  }

  func setProgress(completedUnitCount: Double, totalUnitCount: Double) throws {
    let total = max(totalUnitCount, 0)
    let completed = min(max(completedUnitCount, 0), total)

    lock.lock()
    currentCompletedUnitCount = completed
    currentTotalUnitCount = total
    let task = nativeTask
    lock.unlock()

    if let progress = task?.progress {
      progress.totalUnitCount = Int64(total)
      progress.completedUnitCount = Int64(completed)
    }
    ContinuedTaskStore.shared.update(id: identifier) { record in
      record.completedUnitCount = completed
      record.totalUnitCount = total
    }
  }

  func complete(success: Bool) throws {
    lock.lock()
    guard currentState == .pending || currentState == .running else {
      lock.unlock()
      return
    }
    currentState = .finished
    let task = nativeTask
    nativeTask = nil
    lock.unlock()

    task?.setTaskCompleted(success: success)
    ContinuedTaskStore.shared.update(id: identifier) { record in
      record.state = ContinuedTaskState.finished.stringValue
    }
    ContinuedTaskRegistry.shared.release(identifier: identifier)
  }

  func cancel() throws {
    lock.lock()
    guard currentState == .pending || currentState == .running else {
      lock.unlock()
      return
    }
    let wasPending = currentState == .pending
    let task = nativeTask
    lock.unlock()

    if wasPending {
      // Still queued: withdraw the request before the system launches it.
      BGTaskScheduler.shared.cancel(taskRequestWithIdentifier: identifier)
    }
    task?.setTaskCompleted(success: false)

    stop(
      reason: .appCancelled,
      detail: NativeStopDetail(domain: "BGTaskScheduler", code: nil, name: "cancel")
    )
  }

  func addOnStartListener(listener: @escaping () -> Void) throws -> ListenerSubscription {
    lock.lock()
    let listenerId = nextListenerId
    nextListenerId += 1
    let isAlreadyRunning = currentState == .running
    if !isAlreadyRunning {
      startListeners[listenerId] = listener
    }
    lock.unlock()

    // Replay for a task that started before this listener was added, so there
    // is no race between submitting and subscribing.
    if isAlreadyRunning {
      listener()
      return ListenerSubscription(remove: {})
    }
    return ListenerSubscription(remove: { [weak self] in
      guard let self else { return }
      self.lock.lock()
      self.startListeners.removeValue(forKey: listenerId)
      self.lock.unlock()
    })
  }

  func addOnStopListener(
    listener: @escaping (TaskStopEvent) -> Void
  ) throws -> ListenerSubscription {
    lock.lock()
    let listenerId = nextListenerId
    nextListenerId += 1
    stopListeners[listenerId] = listener
    lock.unlock()

    return ListenerSubscription(remove: { [weak self] in
      guard let self else { return }
      self.lock.lock()
      self.stopListeners.removeValue(forKey: listenerId)
      self.lock.unlock()
    })
  }

  // MARK: - System callbacks

  /// Called from the scheduler's launch handler when the task actually starts.
  func attach(nativeTask task: BGContinuedProcessingTask) {
    lock.lock()
    guard currentState == .pending else {
      lock.unlock()
      task.setTaskCompleted(success: false)
      return
    }
    nativeTask = task
    currentState = .running
    let completed = currentCompletedUnitCount
    let total = currentTotalUnitCount
    let listeners = Array(startListeners.values)
    startListeners.removeAll()
    lock.unlock()

    task.progress.totalUnitCount = Int64(total)
    task.progress.completedUnitCount = Int64(completed)
    task.expirationHandler = { [weak self] in
      // iOS routes both user cancellation from the Live Activity and system
      // expiration through this one handler, with nothing to tell them apart,
      // so this reports the reason it can actually justify.
      self?.stop(
        reason: .expired,
        detail: NativeStopDetail(
          domain: "BGTaskScheduler",
          code: nil,
          name: "expirationHandler"
        )
      )
    }

    ContinuedTaskStore.shared.update(id: identifier) { record in
      record.state = ContinuedTaskState.running.stringValue
    }
    listeners.forEach { $0() }
  }

  private func stop(reason: TaskStopReason, detail: NativeStopDetail) {
    lock.lock()
    guard currentState == .pending || currentState == .running else {
      lock.unlock()
      return
    }
    currentState = .stopped
    nativeTask = nil
    let listeners = Array(stopListeners.values)
    startListeners.removeAll()
    lock.unlock()

    ContinuedTaskStore.shared.update(id: identifier) { record in
      record.state = ContinuedTaskState.stopped.stringValue
      record.stopReason = reason.stringValue
    }
    ContinuedTaskRegistry.shared.release(identifier: identifier)

    let event = TaskStopEvent(taskId: identifier, reason: reason, native: detail)
    listeners.forEach { $0(event) }
  }
}

File: android/src/main/java/com/margelo/nitro/continuedtask/HybridContinuedTask.kt
package com.margelo.nitro.continuedtask

import android.content.Context
import androidx.annotation.Keep
import androidx.work.WorkManager
import com.facebook.proguard.annotations.DoNotStrip
import kotlinx.coroutines.CompletableDeferred
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.launch
import java.util.UUID
import java.util.concurrent.ConcurrentHashMap
import java.util.concurrent.atomic.AtomicInteger

/**
 * A live handle to one WorkManager task running as a foreground service.
 *
 * The handle is created by the manager at submit time and paired with its
 * [ContinuedTaskWorker] through [ContinuedTaskRegistry] once the worker starts.
 */
@DoNotStrip
@Keep
class HybridContinuedTask(
  private val context: Context,
  private val identifier: String,
  private val workId: UUID,
  private val store: ContinuedTaskStore,
  initialTitle: String,
  initialSubtitle: String,
  initialTotalUnitCount: Double,
  internal val androidOptions: AndroidTaskOptions?
) : HybridContinuedTaskSpec() {

  private val scope = CoroutineScope(SupervisorJob() + Dispatchers.Default)
  private val completion = CompletableDeferred<Boolean>()
  private val startListeners = ConcurrentHashMap<Int, () -> Unit>()
  private val stopListeners = ConcurrentHashMap<Int, (TaskStopEvent) -> Unit>()
  private val nextListenerId = AtomicInteger(0)

  private val display = MutableStateFlow(
    TaskDisplayState(initialTitle, initialSubtitle, 0.0, initialTotalUnitCount)
  )

  /** Observed by the worker to keep the notification current. */
  internal val displayState: StateFlow<TaskDisplayState> get() = display

  @Volatile
  private var currentState: ContinuedTaskState = ContinuedTaskState.PENDING

  @Volatile
  private var worker: ContinuedTaskWorker? = null

  override val id: String get() = identifier
  override val title: String get() = display.value.title
  override val subtitle: String get() = display.value.subtitle
  override val state: ContinuedTaskState get() = currentState
  override val completedUnitCount: Double get() = display.value.completedUnitCount
  override val totalUnitCount: Double get() = display.value.totalUnitCount

  override fun updateTitle(title: String, subtitle: String) {
    display.value = display.value.copy(title = title, subtitle = subtitle)
    store.update(identifier) { record ->
      record.title = title
      record.subtitle = subtitle
    }
  }

  override fun setProgress(completedUnitCount: Double, totalUnitCount: Double) {
    val total = maxOf(totalUnitCount, 0.0)
    val completed = minOf(maxOf(completedUnitCount, 0.0), total)
    display.value = display.value.copy(
      completedUnitCount = completed,
      totalUnitCount = total
    )
    store.update(identifier) { record ->
      record.completedUnitCount = completed
      record.totalUnitCount = total
    }
    val activeWorker = worker
    if (activeWorker != null) {
      scope.launch { activeWorker.publishProgress(completed, total) }
    }
  }

  override fun complete(success: Boolean) {
    if (currentState != ContinuedTaskState.PENDING && currentState != ContinuedTaskState.RUNNING) {
      return
    }
    currentState = ContinuedTaskState.FINISHED
    store.update(identifier) { record ->
      record.state = ContinuedTaskRecord.stateName(ContinuedTaskState.FINISHED)
    }
    ContinuedTaskRegistry.release(identifier)
    completion.complete(success)
  }

  override fun cancel() {
    if (currentState != ContinuedTaskState.PENDING && currentState != ContinuedTaskState.RUNNING) {
      return
    }
    // Cancelling the work stops the worker, whose cancellation handler reports
    // STOP_REASON_CANCELLED_BY_APP back through onWorkerStopped.
    WorkManager.getInstance(context).cancelWorkById(workId)
    if (worker == null) {
      // Never started, so no worker will report the stop.
      emitStop(TaskStopReason.APP_CANCELLED, WorkManagerStopReasons.CANCELLED_BY_APP_NAME, null)
    }
  }

  override fun addOnStartListener(listener: () -> Unit): ListenerSubscription {
    if (currentState == ContinuedTaskState.RUNNING) {
      // Replay for a task that started before this listener was added.
      listener()
      return ListenerSubscription(Func_void {})
    }
    val listenerId = nextListenerId.getAndIncrement()
    startListeners[listenerId] = listener
    return ListenerSubscription(Func_void { startListeners.remove(listenerId) })
  }

  override fun addOnStopListener(listener: (event: TaskStopEvent) -> Unit): ListenerSubscription {
    val listenerId = nextListenerId.getAndIncrement()
    stopListeners[listenerId] = listener
    return ListenerSubscription(Func_void { stopListeners.remove(listenerId) })
  }

  // MARK: - Worker callbacks

  internal fun onWorkerStarted(startedWorker: ContinuedTaskWorker) {
    worker = startedWorker
    if (currentState != ContinuedTaskState.PENDING) return
    currentState = ContinuedTaskState.RUNNING
    store.update(identifier) { record ->
      record.state = ContinuedTaskRecord.stateName(ContinuedTaskState.RUNNING)
    }
    val listeners = startListeners.values.toList()
    startListeners.clear()
    listeners.forEach { it() }
  }

  internal suspend fun awaitCompletion(): Boolean = completion.await()

  internal fun onWorkerStopped(stopReason: Int) {
    emitStop(
      TaskStopReasons.fromWorkManagerStopReason(stopReason),
      TaskStopReasons.nameFor(stopReason),
      stopReason.toDouble()
    )
  }

  private fun emitStop(reason: TaskStopReason, name: String, code: Double?) {
    if (currentState != ContinuedTaskState.PENDING && currentState != ContinuedTaskState.RUNNING) {
      return
    }
    currentState = ContinuedTaskState.STOPPED
    store.update(identifier) { record ->
      record.state = ContinuedTaskRecord.stateName(ContinuedTaskState.STOPPED)
      record.stopReason = ContinuedTaskRecord.stopReasonName(reason)
    }
    ContinuedTaskRegistry.release(identifier)
    startListeners.clear()

    val event = TaskStopEvent(
      taskId = identifier,
      reason = reason,
      native = NativeStopDetail(domain = "WorkManager", code = code, name = name)
    )
    val listeners = stopListeners.values.toList()
    listeners.forEach { it(event) }
    completion.complete(false)
  }
}
